Grid/qcd/action/gauge/LLRGaugeAction.h

```diff
@@ -0,0 +1,162 @@
+        return action;
+    };
+    /////////////////////////////////////////////////////////////
+    /// [Deriv]
+    /////////////////////////////////////////////////////////////
+    virtual void deriv(const GaugeField &U, GaugeField &dSdU) {
+        // not optimal implementation FIXME
+        // extend Ta to include Lorentz indexes
+        std::cout << GridLogLLR << B_GREEN   << "(deriv)     LLRGaugeAction"              << C_RESET   << std::endl;
+        std::cout << GridLogLLR << B_MAGENTA << "beta(deriv) LLRGaugeAction      ----->: "<< beta      << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_MAGENTA << "Nc(deriv)   LLRGaugeAction      ----->: "<< RealD(Nc) << C_RESET <<std::endl;
+
+        RealD factor = 0.5  / RealD(Nc);//0.5 * beta / RealD(Nc); //0.5 *
+        GridBase *grid = U.Grid();
+
+        std::cout << GridLogLLR << B_YELLOW   << "factor(deriv) LLRGaugeAction    ----->: "<< factor << C_RESET <<std::endl;
+
+        GaugeLinkField dSdU_mu(grid);
+        std::vector<GaugeLinkField> Umu(Nd, grid);
+
+        for (int mu = 0; mu < Nd; mu++) {
+            Umu[mu] = PeekIndex<LorentzIndex>(U, mu);
+        }
+
+        // Get the average Plaquette to compute S_0
+        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
+        RealD vol = U.Grid()->gSites();
+        // now construct the S_0
+        RealD S0_plaq_ = 6 * vol * plaq;
+        // hard coded value for now TODO: remove this value when properly setup
+        RealD S0_ = p_llrparams_s->S0; //173088.00000;
+
+        std::cout << GridLogLLR << B_RED     << "Vol(deriv)      LLRGaugeAction  ----->: "<< vol               << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_CYAN    << "plaq(deriv)     LLRGaugeAction  ----->: "<< plaq              << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_YELLOW  << "S0_plaq_(deriv) LLRGaugeAction  ----->: "<< S0_plaq_          << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_MAGENTA << "p->S0(deriv)    LLRGaugeAction  ----->: "<< p_llrparams_s->S0 << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_YELLOW  << "p->a(deriv)     LLRGaugeAction  ----->: "<< p_llrparams_s->a  << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_YELLOW  << "p->dS(deriv)    LLRGaugeAction  ----->: "<< p_llrparams_s->dS << C_RESET <<std::endl;
+
+        //GaugeLinkField new_dSdU_mu(grid);
+        RealD action_u = 0.0;
+        action_u = S(U);
+        for (int mu = 0; mu < Nd; mu++) {
+            // Getting the action for when action
+            // Staple in direction mu
+            WilsonLoops<Gimpl>::Staple(dSdU_mu, Umu, mu);
+
+            dSdU_mu *= ( p_llrparams_s->a + (action_u - p_llrparams_s->S0) / (p_llrparams_s->dS * p_llrparams_s->dS) );
+            dSdU_mu = Ta(Umu[mu] * dSdU_mu) * factor;
+
+            PokeIndex<LorentzIndex>(dSdU, dSdU_mu, mu);
+        }
```

Comment:
Similarly to the comment about `S` above, this copy-paste of `WilsonGaugeAction` means that the template parameter is ignored, and this always uses the Wilson action.

Unlike the `S` case, we can't just directly call the `deriv` method of the template action as-is, because we need to do the multiplication by the LLR factor before performing the projection on line 136.

Please can you think about ways around this, that will allow us to correctly template on the gauge action?